pkg/operatorstatus/operatorstatus.go

```diff
@@ -59,44 +58,63 @@ type ClusterOperatorWrapper interface {
 
 func (o *ClusterOperatorHandler) UpdateOperatorStatus(sampleResource *v1alpha1.SamplesResource) error {
+
 	var err error
+
 	switch {
-	case sampleResource.ConditionTrue(v1alpha1.SamplesExist):
-		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionTrue, "Samples exist in the openshift project")
+	// samples exists == true means samples created; versions matching means imagestream imports done
+	case sampleResource.ConditionTrue(v1alpha1.SamplesExist) && sampleResource.Spec.Version == sampleResource.Status.Version:
+		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionTrue, fmt.Sprintf("Samples installed at %s", sampleResource.Status.Version), "")
 	case sampleResource.ConditionFalse(v1alpha1.SamplesExist):
-		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionFalse, "Samples do not exist in the openshift project")
-	default:
-		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionUnknown, "The presence of the samples in the openshift project is unknown")
+	case sampleResource.ConditionUnknown(v1alpha1.SamplesExist):
+		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionFalse, fmt.Sprintf("Samples not yet installed at %s", sampleResource.Spec.Version), "")
 	}
 	if err != nil {
 		return err
 	}
 
+	errorDetails := sampleResource.ConditionsMessages()
 	switch {
-	case sampleResource.ConditionTrue(v1alpha1.ConfigurationValid):
-		err = o.setOperatorStatus(configv1.OperatorFailing, configv1.ConditionFalse, "The samples operator configuration is valid")
+	// errorDetails should have data if config valid false or progress unknown, but just in case
 	case sampleResource.ConditionFalse(v1alpha1.ConfigurationValid):
-		err = o.setOperatorStatus(configv1.OperatorFailing, configv1.ConditionTrue, "The samples operator configuration is invalid")
+	case sampleResource.ConditionUnknown(v1alpha1.ImageChangesInProgress):
+	case len(errorDetails) > 0:
+		err = o.setOperatorStatus(configv1.OperatorFailing, configv1.ConditionTrue, fmt.Sprintf("Unable to install samples at %s: %s", sampleResource.Spec.Version, errorDetails), "")
+		if err != nil {
+			return err
+		}
+		err = o.setOperatorStatus(configv1.OperatorAvailable, configv1.ConditionFalse, fmt.Sprintf("Unable to install samples at %s", sampleResource.Spec.Version), "")
+		if err != nil {
+			return err
+		}
+		briefStateDescription := ""
+		switch {
+		case sampleResource.ConditionFalse(v1alpha1.ConfigurationValid):
+			briefStateDescription = "invalid configuration"
+		case sampleResource.ConditionFalse(v1alpha1.ImportCredentialsExist) && sampleResource.Spec.InstallType == v1alpha1.RHELSamplesDistribution:
+			briefStateDescription = "image pull credentials needed"
+		}
+		// on errors we stop after this setting
+		return o.setOperatorStatus(configv1.OperatorProgressing, configv1.ConditionTrue, fmt.Sprintf("Unable to install samples at %s: %s", sampleResource.Spec.Version, briefStateDescription), "")
```

Comment:
I also think this whole block confuses the issue a bit... you've got a block above that's solely responsible for setting Available, and a block below that's soley responsible for for setting Progressing, but then this block, which seems like it is mostly responsible for setting Failing, also contains logic for setting Available+Progressing in certain cases.

Can this be refactored so that all the logic for setting Progressing is in one place, Available in another, and Failing in a third, rather than intermingling it?

Reply:
ah dammit ... I meant to put a //TODO comment explaining some things around this, but then forgot as I was rushing to get out of the office.

Yeah, while this functionally works, this is not the final form.  It was simply the form with the least amount change.  It leverages the current side effect of the implementation that the `Messages` field is only set on errors.

 There are a couple of probably better choices, but they involved a larger change, and I wanted to discuss and got consensus on which choice before making the investment on the change.

The 3 condition choices are `ConditionTrue`, `ConditionFalse`, and `ConditionUnknown`.  Per the godoc, "ConditionUnknown" means kubernetes can't decide if a resource is in the condition or not.  Whether "errors" should be `ConditionFalse` or `ConditionUnknown` has been a back and forth with me.  The current mixture of use 

One possible choice would be to consider *all* errors, whether we are unsure, or absolutely sure, on the errors, with a `ConditionUnknown` setting.  Using `ConditionFalse` muddies the waters, because distinguishing between error state and uninitialized or pre steady state becomes problematic.  But getting consistent on the condition value, the bloc you mentioned we be refactored along checking for `ConditionUnknown` lines.

Another choice is to explicitly add a "Failing" condition, and use it to map to the failing condition, and redo the block above along those lines.

Thoughts?

Reply:
I think ConditionUnknown and ConditionFalse are orthogonal to errors.  For us it is up to us and what our operator can determine about the state of the system.

If we are sure the sample exist, available=true.  if we're sure they don't exist (because we know we failed to create them for some reason or are blocked from creating them, or deleted them), available=false.  If we got an error part way into creating or deleting them, then we're in muddy waters where we should either be setting it to false (if we're confident we didn't manage to create everything) or unknown (if we really can't be sure if the creation/deletion went through).

We probably need to create a truth table that goes through the various samplesresource condition combinations, and then what the clusteroperator conditions should be given those samplesresource conditions.  (which is basically what you've constructed in your switch statement, but perhaps it we write it out formally somewhere things will become clearer).

> Whether "errors" should be ConditionFalse or ConditionUnknown has been a back and forth with me

just to elaborate on this point:  an error could result in a conditiontrue also (such as in the case of the Failing condition).  So i view the error as "additional information about why we are in the state we are reporting", the error itself does not determine what state we are in, it provides context to why we are reporting that state.

Reply:
> Next, samples exists==true != available==true.

right, i forgot about the need to actually do the import/have the import succeed.  my bad. I don't think it alters my larger view on how to structure/determine the states though.


> I'm just don't think having the presence of a non-empty message on the existing conditions on the samplesresource as a good input in all cases for the truth table to determine things (it can in a case or two). Too fragile. 

agree.

> Solely using conditiontrue/false/unknown on the existing set of conditions in the samplesresource is insufficient as well, where our statements above capture that in my opinion.

hm, this one i'm not following.  Can you give me a specific example where a given set of conditions exists on the samplesresource, but those conditions (without the associated reason/message field) do not uniquely determine the appropriate set of clusteroperator condition values?  (setting aside what the reason/message should be on the clusteroperator conditions).  (You say it's above and it may be but i'm having trouble extracting it from the discussion at the moment).

Assuming their is such a case, i might be inclined to think that indicates a need for additional conditions on the samplesresource, but i guess i'll have to see what the scenario is first.